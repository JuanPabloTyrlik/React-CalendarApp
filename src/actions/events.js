import { TYPES } from '../types/types';

export const eventAddNew = (event) => {
    return {
        type: TYPES.EVENT_ADD_NEW,
        payload: event,
    };
};

export const eventSetActive = (event) => {
    return {
        type: TYPES.EVENT_SET_ACTIVE,
        payload: event,
    };
};
