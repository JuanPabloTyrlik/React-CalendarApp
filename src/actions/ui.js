import { TYPES } from '../types/types';

export const uiModalOpen = () => {
    return {
        type: TYPES.UI_OPEN_MODAL,
        payload: null,
    };
};

export const uiModalClose = () => {
    return {
        type: TYPES.UI_CLOSE_MODAL,
        payload: null,
    };
};
