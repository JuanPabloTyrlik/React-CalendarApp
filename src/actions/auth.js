import Swal from 'sweetalert2';
import { fetchWithoutToken, fetchWithToken } from '../helpers/fetch';
import { TYPES } from '../types/types';
import { eventClearAll } from './events';

// Async Actions
export const startLogin = (email, password) => async (dispatch) => {
    const resp = await fetchWithoutToken(
        'auth/login',
        { email, password },
        'POST'
    );
    const body = await resp.json();
    if (body.ok) {
        localStorage.setItem('token', body.token);
        localStorage.setItem('token-init-date', new Date().getTime());
        dispatch(
            login({
                uid: body.uid,
                name: body.name,
            })
        );
    } else {
        if (body.errors) {
            body.message = body.errors.reduce(
                (a, v) => (a += '' + v.msg + '<br/>'),
                ''
            );
        }
        Swal.fire('Error', body.message, 'error');
    }
};

export const startRegister = (name, email, password) => async (dispatch) => {
    const resp = await fetchWithoutToken(
        'auth/register',
        { name, email, password },
        'POST'
    );
    const body = await resp.json();
    if (body.ok) {
        localStorage.setItem('token', body.token);
        localStorage.setItem('token-init-date', new Date().getTime());
        dispatch(
            login({
                uid: body.uid,
                name: body.name,
            })
        );
    } else {
        if (body.errors) {
            body.message = body.errors.reduce(
                (a, v) => (a += '' + v.msg + '<br/>'),
                ''
            );
        }
        Swal.fire('Error', body.message, 'error');
    }
};

export const startChecking = () => async (dispatch) => {
    const resp = await fetchWithToken('auth/token');
    const body = await resp.json();
    if (body.ok) {
        localStorage.setItem('token', body.token);
        localStorage.setItem('token-init-date', new Date().getTime());
        dispatch(
            login({
                uid: body.uid,
                name: body.name,
            })
        );
    } else {
        dispatch(checkingFinish());
    }
};

export const startLogout = () => (dispatch) => {
    localStorage.clear();
    dispatch(logout());
    dispatch(eventClearAll());
};

// Sync Actions
export const login = (user) => {
    return {
        type: TYPES.AUTH_LOGIN,
        payload: user,
    };
};

export const logout = () => {
    return {
        type: TYPES.AUTH_LOGOUT,
        payload: null,
    };
};

export const checkingFinish = () => {
    return {
        type: TYPES.AUTH_CHECKING_FINISH,
        payload: null,
    };
};
