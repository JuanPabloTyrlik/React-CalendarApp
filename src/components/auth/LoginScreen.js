import React, { useState } from 'react';
import validator from 'validator';
import { useDispatch } from 'react-redux';
import { startLogin, startRegister } from '../../actions/auth';
import { useForm } from '../../hooks/useForm';
import './login.css';

export const LoginScreen = () => {
    const dispatch = useDispatch();

    const [errors, setErrors] = useState({
        Lemail: false,
        Lpassword: false,
        Rname: false,
        Remail: false,
        Rpassword: false,
        Rpassword2: false,
    });

    const [loginForm, handleLoginChange] = useForm({
        Lemail: '',
        Lpassword: '',
    });

    const [registerForm, handleRegisterChange] = useForm({
        Rname: '',
        Remail: '',
        Rpassword: '',
        Rpassword2: '',
    });

    const { Lemail, Lpassword } = loginForm;
    const { Rname, Remail, Rpassword, Rpassword2 } = registerForm;

    const handleLogin = (e) => {
        e.preventDefault();
        isValidLoginForm();
        dispatch(startLogin(Lemail, Lpassword));
    };

    const handleRegister = (e) => {
        e.preventDefault();
        isValidRegisterForm();
        dispatch(startRegister(Rname, Remail, Rpassword));
    };

    const isValidLoginForm = () => {
        if (!validator.isEmail(Lemail)) {
            setErrors({ ...errors, Lemail: true });
            return false;
        }
        if (Lpassword.trim().length < 8) {
            setErrors({ ...errors, Lpassword: true });
            return false;
        }
        setErrors({
            Lemail: false,
            Lpassword: false,
            Rname: false,
            Remail: false,
            Rpassword: false,
            Rpassword2: false,
        });
        return true;
    };

    const isValidRegisterForm = () => {
        if (/^\s*$/.test(Rname)) {
            setErrors({ ...errors, Rname: true });
            return false;
        }
        if (!validator.isEmail(Remail)) {
            setErrors({ ...errors, Remail: true });
            return false;
        }
        if (Rpassword.trim().length < 8 || Rpassword.trim().length > 32) {
            setErrors({ ...errors, Rpassword: true });
            return false;
        }
        if (!/[0-9]/.test(Rpassword)) {
            setErrors({ ...errors, Rpassword: true });
            return false;
        }
        if (!/[a-z]/.test(Rpassword)) {
            setErrors({ ...errors, Rpassword: true });
            return false;
        }
        if (!/[A-Z]/.test(Rpassword)) {
            setErrors({ ...errors, Rpassword: true });
            return false;
        }
        if (!/[*.!@$%^&(){}[\]:;,.?/\\~_+\-=|]/.test(Rpassword)) {
            setErrors({ ...errors, Rpassword: true });
            return false;
        }
        if (Rpassword !== Rpassword2) {
            setErrors({ ...errors, Rpassword: true });
            setErrors({ ...errors, Rpassword2: true });
            return false;
        }
        setErrors({
            Lemail: false,
            Lpassword: false,
            Rname: false,
            Remail: false,
            Rpassword: false,
            Rpassword2: false,
        });
        return true;
    };

    return (
        <div className="container login-container">
            <div className="row">
                <div className="col-md-6 login-form-1">
                    <h3>Ingreso</h3>
                    <form onSubmit={handleLogin}>
                        <div className="form-group">
                            <input
                                type="text"
                                className={
                                    errors.Lemail
                                        ? 'form-control is-invalid'
                                        : 'form-control'
                                }
                                placeholder="Correo"
                                name="Lemail"
                                value={Lemail}
                                onChange={handleLoginChange}
                            />
                        </div>
                        <div className="form-group">
                            <input
                                type="password"
                                className={
                                    errors.Lpassword
                                        ? 'form-control is-invalid'
                                        : 'form-control'
                                }
                                placeholder="Contraseña"
                                name="Lpassword"
                                value={Lpassword}
                                onChange={handleLoginChange}
                            />
                        </div>
                        <div className="form-group">
                            <input
                                type="submit"
                                className="btnSubmit"
                                value="Login"
                            />
                        </div>
                    </form>
                </div>

                <div className="col-md-6 login-form-2">
                    <h3>Registro</h3>
                    <form onSubmit={handleRegister}>
                        <div className="form-group">
                            <input
                                type="text"
                                className={
                                    errors.Rname
                                        ? 'form-control is-invalid'
                                        : 'form-control'
                                }
                                placeholder="Nombre"
                                name="Rname"
                                value={Rname}
                                onChange={handleRegisterChange}
                            />
                        </div>
                        <div className="form-group">
                            <input
                                type="email"
                                className={
                                    errors.Remail
                                        ? 'form-control is-invalid'
                                        : 'form-control'
                                }
                                placeholder="Correo"
                                name="Remail"
                                value={Remail}
                                onChange={handleRegisterChange}
                            />
                        </div>
                        <div className="form-group">
                            <input
                                type="password"
                                className={
                                    errors.Rpassword
                                        ? 'form-control is-invalid'
                                        : 'form-control'
                                }
                                placeholder="Contraseña"
                                name="Rpassword"
                                value={Rpassword}
                                onChange={handleRegisterChange}
                            />
                        </div>

                        <div className="form-group">
                            <input
                                type="password"
                                className={
                                    errors.Rpassword2
                                        ? 'form-control is-invalid'
                                        : 'form-control'
                                }
                                placeholder="Repita la contraseña"
                                name="Rpassword2"
                                value={Rpassword2}
                                onChange={handleRegisterChange}
                            />
                        </div>

                        <div className="form-group">
                            <input
                                type="submit"
                                className="btnSubmit"
                                value="Crear cuenta"
                            />
                        </div>
                    </form>
                </div>
            </div>
        </div>
    );
};
