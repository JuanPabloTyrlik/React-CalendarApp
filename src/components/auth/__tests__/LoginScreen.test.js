import React from 'react';
import { mount } from 'enzyme';
import { Provider } from 'react-redux';
import configureStore from 'redux-mock-store';
import thunk from 'redux-thunk';
import { LoginScreen } from '../LoginScreen';
import { act } from '@testing-library/react';
import { startLogin, startRegister } from '../../../actions/auth';
import Swal from 'sweetalert2';

const middlewares = [thunk];
const mockStore = configureStore(middlewares);

const store = mockStore({});
store.dispatch = jest.fn();
jest.mock('../../../actions/auth', () => ({
    startLogin: jest.fn(),
    startRegister: jest.fn(),
}));

jest.mock('sweetalert2', () => ({
    fire: jest.fn(),
}));

const wrapper = mount(
    <Provider store={store}>
        <LoginScreen />
    </Provider>
);

describe('Tests on <LoginScreen />', () => {
    beforeEach(() => {
        store.clearActions();
        jest.clearAllMocks();
    });
    test('should render correctly', () => {
        expect(wrapper).toMatchSnapshot();
    });
    test('should log an user', async () => {
        const { email, password } = JSON.parse(
            process.env.REACT_APP_API_TEST_USER
        );
        wrapper.find('input[name="Lemail"]').simulate('change', {
            target: {
                name: 'Lemail',
                value: email,
            },
        });
        wrapper.find('input[name="Lpassword"]').simulate('change', {
            target: {
                name: 'Lpassword',
                value: password,
            },
        });
        act(() => {
            wrapper.find('form').at(0).prop('onSubmit')({
                preventDefault() {},
            });
        });
        expect(store.dispatch).toHaveBeenCalled();
        expect(startLogin).toHaveBeenCalledWith(email, password);
    });
    test('should show error messages and register once resolved', async () => {
        act(() => {
            wrapper.find('form').at(1).prop('onSubmit')({
                preventDefault() {},
            });
        });
        expect(Swal.fire).toHaveBeenCalledWith(
            'Error',
            'Name cannot be empty',
            'error'
        );
        wrapper.find('input[name="Rname"]').simulate('change', {
            target: {
                name: 'Rname',
                value: 'TestUser',
            },
        });
        act(() => {
            wrapper.find('form').at(1).prop('onSubmit')({
                preventDefault() {},
            });
        });
        expect(Swal.fire).toHaveBeenCalledWith(
            'Error',
            'Email must be valid',
            'error'
        );
        wrapper.find('input[name="Remail"]').simulate('change', {
            target: {
                name: 'Remail',
                value: 'valid@email.com',
            },
        });
        act(() => {
            wrapper.find('form').at(1).prop('onSubmit')({
                preventDefault() {},
            });
        });
        expect(Swal.fire).toHaveBeenCalledWith(
            'Error',
            'Password cannot be empty',
            'error'
        );
        wrapper.find('input[name="Rpassword"]').simulate('change', {
            target: {
                name: 'Rpassword',
                value: 'ThisIs-A-Password-1',
            },
        });
        act(() => {
            wrapper.find('form').at(1).prop('onSubmit')({
                preventDefault() {},
            });
        });
        expect(Swal.fire).toHaveBeenCalledWith(
            'Error',
            "Passwords don't match",
            'error'
        );
        wrapper.find('input[name="Rpassword2"]').simulate('change', {
            target: {
                name: 'Rpassword2',
                value: 'ThisIs-A-Password-1',
            },
        });
        act(() => {
            wrapper.find('form').at(1).prop('onSubmit')({
                preventDefault() {},
            });
        });
        expect(store.dispatch).toHaveBeenCalled();
        expect(startRegister).toHaveBeenCalledWith(
            'TestUser',
            'valid@email.com',
            'ThisIs-A-Password-1'
        );
    });
});
