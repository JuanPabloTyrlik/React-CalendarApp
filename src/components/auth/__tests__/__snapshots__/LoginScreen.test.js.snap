// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tests on <LoginScreen /> should render correctly 1`] = `
<div
  className="container login-container"
>
  <div
    className="row"
  >
    <div
      className="col-md-6 login-form-1"
    >
      <h3>
        Ingreso
      </h3>
      <form
        onSubmit={[Function]}
      >
        <div
          className="form-group"
        >
          <input
            className="form-control"
            name="Lemail"
            onChange={[Function]}
            placeholder="Correo"
            type="text"
            value=""
          />
        </div>
        <div
          className="form-group"
        >
          <input
            className="form-control"
            name="Lpassword"
            onChange={[Function]}
            placeholder="Contraseña"
            type="password"
            value=""
          />
        </div>
        <div
          className="form-group"
        >
          <input
            className="btnSubmit"
            type="submit"
            value="Login"
          />
        </div>
      </form>
    </div>
    <div
      className="col-md-6 login-form-2"
    >
      <h3>
        Registro
      </h3>
      <form
        onSubmit={[Function]}
      >
        <div
          className="form-group"
        >
          <input
            className="form-control"
            name="Rname"
            onChange={[Function]}
            placeholder="Nombre"
            type="text"
            value=""
          />
        </div>
        <div
          className="form-group"
        >
          <input
            className="form-control"
            name="Remail"
            onChange={[Function]}
            placeholder="Correo"
            type="email"
            value=""
          />
        </div>
        <div
          className="form-group"
        >
          <input
            className="form-control"
            name="Rpassword"
            onChange={[Function]}
            placeholder="Contraseña"
            type="password"
            value=""
          />
        </div>
        <div
          className="form-group"
        >
          <input
            className="form-control"
            name="Rpassword2"
            onChange={[Function]}
            placeholder="Repita la contraseña"
            type="password"
            value=""
          />
        </div>
        <div
          className="form-group"
        >
          <input
            className="btnSubmit"
            type="submit"
            value="Crear cuenta"
          />
        </div>
      </form>
    </div>
  </div>
</div>
`;
